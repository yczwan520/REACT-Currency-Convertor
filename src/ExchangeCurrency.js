import React, { useState, useEffect } from "react";
import { message, Select, Spin } from "antd";
import moment from "moment";
// eslint-disable-next-line
import "antd/dist/antd.css"; //

const { Option } = Select;
const ExchangeCurrency = (props) => {
  const [currency, setCurrency] = useState("USD");
  const [rate, setRate] = useState("");
  const [date, setDate] = useState(moment().format("YYYY-MM-DD"));
  const [amount, setAmount] = useState(1);
  const [amountInFromCurrency, setAmountInFromCurrency] = useState(true);
  const [isLoading, setLoading] = useState(true);

  let toAmount, fromAmount;
  if (amountInFromCurrency) {
    fromAmount = amount;
    toAmount = (amount * rate).toFixed(4);
  } else {
    toAmount = amount;
    fromAmount = (amount / rate).toFixed(4);
  }

  const fetchRate = () => {
    setLoading(true);
    const res = fetch(
      `https://www.bankofcanada.ca/valet/observations/FXCAD${currency}?start_date=${
        props.date ? props.date : date
      }&end_date=${props.date ? props.date : date}`
    )
      .then((response) => response.json())
      .then((data) => {
        if (data.observations.length) {
          setRate(data.observations[0][`FXCAD${currency}`].v);
        } else {
          message.info("There is no exchange rate on this date!");
          setRate("");
        }
        setLoading(false);
      })
      .catch((error) => console.log(error));
  };

  useEffect(() => {
    fetchRate();
  }, [currency, props.date]);

  function handleFromAmountChange(e) {
    setAmount(e.target.value);
    setAmountInFromCurrency(true);
  }

  function handleToAmountChange(e) {
    setAmount(e.target.value);
    setAmountInFromCurrency(false);
  }

  return (
    <div>
      <div>
        {isLoading && <Spin />}
        <form className="form1">
          <label className="input">
            Amount of Foreign Currency:
            <input
              type="number"
              name="foreignAmount"
              value={toAmount}
              onChange={handleToAmountChange}
            />
          </label>
          <Select
            currency
            defaultValue="USD"
            style={{
              width: 120
            }}
            onChange={(value) => setCurrency(value)}
          >
            <Option value="USD">USD</Option>
            <Option value="EUR">EUR</Option>
            <Option value="JPY">JPY</Option>
            <Option value="GBP">GBP</Option>
            <Option value="AUD">AUD</Option>
            <Option value="CHF">CHF</Option>
            <Option value="CNY">CNY</Option>
            <Option value="HKD">HKD</Option>
            <Option value="INR">INR</Option>
            <Option value="IDR">IDR</Option>
            <Option value="NZD">NZD</Option>
            <Option value="NOK">NOK</Option>
          </Select>
        </form>
        <form className="form2">
          <label className="input">
            Amount of Canadian Dollar:
            <input
              type="number"
              name="cadAmount"
              value={fromAmount}
              onChange={handleFromAmountChange}
            />
          </label>

          <Select
            defaultValue="CAD"
            style={{
              width: 120
            }}
            disabled
          >
            <Option value="CAD">CAD</Option>
          </Select>
        </form>
      </div>
      <div>
        <p>Today's exchange rate is ${rate}</p>
      </div>
    </div>
  );
};
export default ExchangeCurrency;
