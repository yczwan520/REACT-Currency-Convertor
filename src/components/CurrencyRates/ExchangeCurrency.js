import React, { useState, useEffect } from "react";
import { message, Select, Spin } from "antd";
import moment from "moment";
import style from "./ExchangeCurrency.module.scss";
// eslint-disable-next-line
import { DatePicker, Space } from "antd";
import { Input } from "antd";
import "antd/dist/antd.css"; //

const { Option } = Select;
const ExchangeCurrency = (props) => {
  const [currency, setCurrency] = useState("USD");
  const [rate, setRate] = useState("");
  const [date, setDate] = useState("");
  const [amount, setAmount] = useState(1);
  const [amountInFromCurrency, setAmountInFromCurrency] = useState(true);
  const [isLoading, setLoading] = useState(true);

  const handleDateChange = (date) => {
    setDate(moment(date).format("YYYY-MM-DD"));
  };

  let toAmount, fromAmount;
  if (amountInFromCurrency) {
    fromAmount = amount;
    toAmount = (amount * rate).toFixed(4);
  } else {
    toAmount = amount;
    fromAmount = (amount / rate).toFixed(4);
  }

  const fetchRate = () => {
    setLoading(true);
    const res = fetch(
      `https://www.bankofcanada.ca/valet/observations/FXCAD${currency}?start_date=${
        props.date ? props.date : date
      }&end_date=${props.date ? props.date : date}`
    )
      .then((response) => response.json())
      .then((data) => {
        if (data.observations.length) {
          setRate(data.observations[0][`FXCAD${currency}`].v);
        } else {
          message.info("There is no exchange rate on this date!");
          setRate("");
        }
        setLoading(false);
      })
      .catch((error) => console.log(error));
  };

  useEffect(() => {
    fetchRate();
  }, [currency, date]);

  function handleFromAmountChange(e) {
    setAmount(e.target.value);
    setAmountInFromCurrency(true);
  }

  function handleToAmountChange(e) {
    setAmount(e.target.value);
    setAmountInFromCurrency(false);
  }

  return (
    <div className={style.convertor}>
      <div>
        <h1>Currency Convertor</h1>
        <div class={style.date_picker}>
          <p className="date">Date</p>
          <DatePicker onChange={(date) => handleDateChange(date)} />
        </div>
        {isLoading && <Spin />}
        <div className={style.exchange_input}>
          <label className="input">
            <p>Amount of Foreign Currency</p>
            <Input
              type="number"
              placeholder="Type amount"
              value={toAmount}
              onChange={handleToAmountChange}
            />
          </label>
          <Select
            currency
            defaultValue="USD"
            style={{
              width: 120,
            }}
            onChange={(value) => setCurrency(value)}
          >
            <Option value="USD">USD</Option>
            <Option value="EUR">EUR</Option>
            <Option value="JPY">JPY</Option>
            <Option value="GBP">GBP</Option>
            <Option value="AUD">AUD</Option>
            <Option value="CHF">CHF</Option>
            <Option value="CNY">CNY</Option>
            <Option value="HKD">HKD</Option>
            <Option value="INR">INR</Option>
            <Option value="IDR">IDR</Option>
            <Option value="NZD">NZD</Option>
            <Option value="NOK">NOK</Option>
          </Select>
        </div>
        <div className={style.exchange_input}>
          <label className="input">
            <p>Amount of Canadian Dollar</p>
            <Input
              type="number"
              placeholder="type amount"
              value={fromAmount}
              onChange={handleFromAmountChange}
            />
          </label>
          <Select
            defaultValue="CAD"
            style={{
              width: 120,
            }}
            disabled
          >
            <Option value="CAD">CAD</Option>
          </Select>
        </div>
      </div>
      <div className={style.exchange_rate}>
        <p>Today's exchange rate is ${rate}</p>
      </div>
    </div>
  );
};
export default ExchangeCurrency;
