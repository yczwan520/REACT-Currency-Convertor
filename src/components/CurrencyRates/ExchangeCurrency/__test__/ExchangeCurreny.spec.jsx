import { render, screen, waitFor } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import ExchangeCurrency from "../ExchangeCurrency";
import "@testing-library/jest-dom";

describe("ExchangeCurrency page", () => {
  const pageInfo1 = "Amount of Foreign Currency";
  const pageInfo2 = "Amount of Canadian Dollar";
  const pageInfo3 = "Today's exchange rate is $";

  it("renders the ExchangeCurrency page", () => {
    const { container } = render(
      <MemoryRouter>
        <ExchangeCurrency />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
    expect(
      screen.getByText(pageInfo1, pageInfo2, pageInfo3)
    ).toBeInTheDocument();
    expect(
      screen.getByRole("spinbutton", { name: "Amount of Foreign Currency" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("spinbutton", { name: "Amount of Canadian Dollar" })
    ).toHaveValue(1);
    expect(screen.getByRole("img", { name: "calendar" })).toBeInTheDocument();
    expect(screen.getByRole("textbox", { name: "" })).toBeInTheDocument();
  });
});
