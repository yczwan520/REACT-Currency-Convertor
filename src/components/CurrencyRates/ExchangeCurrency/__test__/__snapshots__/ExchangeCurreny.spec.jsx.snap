// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeCurrency page renders the ExchangeCurrency page 1`] = `
<div>
  <div
    class="convertor"
  >
    <div>
      <h1>
        Currency Convertor
      </h1>
      <div
        class="date_picker"
      >
        <p
          class="date"
        >
          Date
        </p>
        <div
          class="ant-picker"
        >
          <div
            class="ant-picker-input"
          >
            <input
              autocomplete="off"
              placeholder="Select date"
              readonly=""
              size="12"
              title=""
              value=""
            />
            <span
              class="ant-picker-suffix"
            >
              <span
                aria-label="calendar"
                class="anticon anticon-calendar"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="calendar"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="ant-spin ant-spin-spinning"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </div>
      <div
        class="exchange_input"
      >
        <label
          class="input"
        >
          <p>
            Amount of Foreign Currency
          </p>
          <input
            class="ant-input"
            placeholder="Type amount"
            type="number"
            value="0.0000"
          />
        </label>
        <div
          class="ant-select ant-select-single ant-select-show-arrow"
          style="width: 120px;"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-search"
            >
              <input
                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="ant-select-selection-search-input"
                id="rc_select_TEST_OR_SSR"
                readonly=""
                role="combobox"
                style="opacity: 0;"
                type="search"
                unselectable="on"
                value=""
              />
            </span>
            <span
              class="ant-select-selection-item"
              title="USD"
            >
              USD
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="down"
              class="anticon anticon-down ant-select-suffix"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
      <div
        class="exchange_input"
      >
        <label
          class="input"
        >
          <p>
            Amount of Canadian Dollar
          </p>
          <input
            class="ant-input"
            placeholder="type amount"
            type="number"
            value="1"
          />
        </label>
        <div
          class="ant-select ant-select-single ant-select-show-arrow ant-select-disabled"
          style="width: 120px;"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-search"
            >
              <input
                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="ant-select-selection-search-input"
                disabled=""
                id="rc_select_TEST_OR_SSR"
                readonly=""
                role="combobox"
                style="opacity: 0;"
                type="search"
                unselectable="on"
                value=""
              />
            </span>
            <span
              class="ant-select-selection-item"
              title="CAD"
            >
              CAD
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="down"
              class="anticon anticon-down ant-select-suffix"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="exchange_rate"
    >
      <p>
        Today's exchange rate is $
      </p>
    </div>
  </div>
</div>
`;
